File: cpp/955_min_deletion_size/min_deletion_size.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int minDeletionSize(vector<string>& strs) {
    // cuts[i] is True : we don't need to check col[i] <= col[i+1]
    vector<bool> cuts(strs.size() - 1, false);
    int ans = 0;
    for (size_t j = 0; j < strs[0].size(); j++) {
      bool canKeep = true;
      for (size_t i = 0; i < strs.size() - 1; i++) {
        if (!cuts[i] && strs[i][j] > strs[i + 1][j]) {
          canKeep = false;
          break;
        }
      }
      if (canKeep) {
        for (size_t i = 0; i < strs.size() - 1; i++) {
          if (strs[i][j] < strs[i + 1][j]) {
            cuts[i] = true;
          }
        }
      } else {
        ans++;
      }
    }
    return ans;
  }
};

int main() {
  vector<tuple<vector<string>, int>> tests{
      {{"ca", "bb", "ac"}, 1},
      {{"xc", "yb", "za"}, 0},
      {{"zyx", "wvu", "tsr"}, 3},
  };

  for (auto& [strs, ans] : tests) {
    assert(Solution().minDeletionSize(strs) == ans);
  }
}

File: rust/955_min_deletion_size/min_deletion_size.rs
struct Solution;

impl Solution {
    pub fn min_deletion_size(strs: Vec<String>) -> i32 {
        // cuts[i] is True : we don't need to check col[i] <= col[i+1]
        let mut cuts = vec![false; strs.len() - 1];
        let mut ans = 0;

        for j in 0..strs[0].len() {
            let mut can_keep = true;
            for i in 0..strs.len() - 1 {
                if !cuts[i] {
                    let char_i = strs[i].chars().nth(j).unwrap();
                    let char_i1 = strs[i + 1].chars().nth(j).unwrap();
                    if char_i > char_i1 {
                        can_keep = false;
                        break;
                    }
                }
            }
            if can_keep {
                for i in 0..strs.len() - 1 {
                    let char_i = strs[i].chars().nth(j).unwrap();
                    let char_i1 = strs[i + 1].chars().nth(j).unwrap();
                    if char_i < char_i1 {
                        cuts[i] = true;
                    }
                }
            } else {
                ans += 1;
            }
        }

        ans
    }
}

fn main() {
    let tests = vec![
        (vec!["ca", "bb", "ac"], 1),
        (vec!["xc", "yb", "za"], 0),
        (vec!["zyx", "wvu", "tsr"], 3),
    ];

    for (strs, ans) in tests {
        assert_eq!(
            Solution::min_deletion_size(strs.iter().map(|s| s.to_string()).collect()),
            ans
        );
    }
}
